### PR TITLE
Refactor testutil

### PR DESCRIPTION
**Pull request by levitte**
Refactor the testutil framework to build as a static library.  This has the benefit that we no longer need to include `test_main.h` / `test_main_custom.h` or link with the corresponding object files.  All that's needed is to include `testutil.h` and link with `libtestutil.a`.

This also makes testutil use BIO for input/output, which means it will work even though stdio is disabled.  `test/bioprinttest.c` is made an exception that outputs directly to file descriptors.

As a side effect, we also get the possibility to build static-only libraries.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Currently in WIP, 'cause I haven't verified that the static-only library build works properly on Windows or VMS yet.

**Comment by dot-asm** [comment]
Red cross from appveoyr was discussed in #3220. [BTW, ping on that, it doesn't clash with this.] You can't switch to BIO_printf and keep OSSLzu.

**Comment by dot-asm** [comment]
> You can't switch to BIO_printf and keep OSSLzu.

To clarify. It's not like there is *no* VC CRT that doesn't understand say `llu`, bit *all* of them do understand `I64u`. Though there is certain ambiguity here, because complete question in the specific context is if there is *64-bit* VC CRT that doesn't understand `llu`. Since I don't know the answer I settled for known-to-work option.

**Comment by levitte** [comment]
Thanks, I'll fix that along with the VMS fixes

**Comment by levitte** [comment]
A number of fix commits added.

I'm actually unsure the strategy with `bioprinttest.c` is sane.  Will we encounter problems with `vsnprintf` (wasn't `BIO_vsnprintf` added because `vsnprintf` doesn't exist everywhere?)?  Why do we want to avoid using BIO for the TEST_* printouts?  It doesn't really seem to me that it's worth the while to avoid it, really...  or at least to avoid using `BIO_vsnprintf`, specifically in a test designed to test it anyway...

**Comment by levitte** [comment]
No longer in WIP, now it's just about working out the kinks...

File: test/bioprinttest.c [hunk 1]
```diff
@@ -11,8 +11,12 @@
 
 #include <stdio.h>
 #include <string.h>
-#include <openssl/opensslconf.h>
-#include OPENSSL_UNISTD
+#include <openssl/e_os2.h>
+#ifdef OPENSSL_SYS_WINDOWS
+# include <winsock.h>
```
Comment (dot-asm) [inline test/bioprinttest.c]: What do you need winsock.h for?

Comment (levitte) [inline test/bioprinttest.c]: `write()`...  that's why I need `unistd.h`, which doesn't exist on Windows.  I had a look at `test/ssltest_old.c` to see how it was done there...

**Comment by dot-asm** [comment]
> I'm actually unsure the strategy with bioprinttest.c is sane.

What strategy? I don't feel that there was any strategy. I said that *if* testutil.c uses BIO_printf, *then* we are better off if bioprinttest.c is not using testutil.c and is executed at level 01. Rich said he is against testutil using BIO_printf. And you said that bioprinttest.c should execute on level 04. It doesn't sound like strategy at all. It sounds more like we are not listening to each other. And I'd argue that the current appveyor result is reflection of that.

File: test/bioprinttest.c (hunk 1 above)
Comment (dot-asm) [inline test/bioprinttest.c]: I'd argue that this is unjustified. There is no reason for test program to "step outside" POSIX. write(2) is not POSIX and shouldn't be used.

Comment (dot-asm) [inline test/bioprinttest.c]: > I'd argue that this is unjustified. There is no reason for test program to "step outside" POSIX. write(2) is not POSIX and shouldn't be used.

I meant standard C library as specified by C language standard, not POSIX. write(2) is part of POSIX, but not standard C library. Sorry about confusion...

Comment (dot-asm) [inline test/bioprinttest.c]: Besides, MSVC defines write in io.h, and preferred way to refer to it is _write. And I don't see no write in winsock.h. WinSock interface doesn't have no write, it has send/recv, you can also use WriteFile on socket, but you can't use write with socket. Because it accepts _fileno(fp), which has *nothing* to do with kind of descriptors used by send/recv or WriteFile.

**Comment by levitte** [comment]
> What strategy? I don't feel that there was any strategy. I said that *if* testutil.c uses BIO_printf, *then* we are better off if bioprinttest.c is not using testutil.c and is executed at level 01. Rich said he is against testutil using BIO_printf. And you said that bioprinttest.c should execute on level 04. It doesn't sound like strategy at all. It sounds more like we are not listening to each other. And I'd argue that the current appveyor result is reflection of that.

You forget @t-j-h, who asked if we want tests to run at all with stdio disabled...  this contradicts the desire to avoid BIO_printf() in testutil...

I'm actually trying to make some kind of sense of it all.

(re 01 vs 04, I'm not sure I see that as an issue...  the 04 group (not *level*, don't let yourself get fooled that there are any *levels*) tests essential APIs, i.e. APIs used by the rest of the code, the 01 group is about more fundamental things)

File: test/bioprinttest.c (hunk 1 above)
Comment (dot-asm) [inline test/bioprinttest.c]: > I'd argue that this is unjustified.

fwrite if you have to...

**Comment by dot-asm** [comment]
> You forget @t-j-h, who asked if we want tests to run at all with stdio disabled...

Right, sorry. Yet I don't feel that question was answered. Well, one can argue that this request is an suggestion of answer... But I still don't feel that there was strategy... At least I wouldn't call it one...

**Comment by levitte** [comment]
I took it by email with the three of you...  it's too hard to follow a discussion that's spread over several PRs

File: test/bioprinttest.c (hunk 1 above)
Comment (levitte) [inline test/bioprinttest.c]: So if we have configured `no-stdio`, we don't run bioprinttest at all?

I'm pausing for now, but am seriously considering undoing the BIO_printf avoiding stuff in this file.

**Comment by levitte** [comment]
I assume, btw, that the sudden appveyor faults is due to using BIO_printf in testutil, correct?  From what I can see, it would help to include `openssl/crypto.h` somewhere, probably in `test/testutil/basic_output.c`...  am I on the right track?

File: test/bioprinttest.c (hunk 1 above)
Comment (dot-asm) [inline test/bioprinttest.c]: > So if we have configured no-stdio, we don't run bioprinttest at all?

There is certain ambiguity in terminology. When we refer to BIO_printf we actually rather refer to its core part, more specifically its ability to produce formatted strings. So that when we say BIO_printf we probably rather mean _dopr. But _dopr actually can be tested with BIO_snprintf and memcmp, isn't it? And that's what bioprinttest.c does in its core. So that if things are so bad, that you don't have stdio and exit code is the only way to convey information, you still can perform meaningful test. It's meaningful, because if things are that bad that there is no stdio, you ought to have some lower-level debugging facility, e.g. through JTAG, and should be able to debug in case failure is returned.

**Comment by levitte** [comment]
That wasn't it...  I finally figured out that the only reason we're seeing OPENSSL_Applink issues on Appveyor is that I chose to use BIO_new_fd()...  so basically, it means that if I want to avoid that, I either need to link all test programs with `$target{apps_aux_src}` or use stdio...

**Comment by levitte** [comment]
I put this back in WIP, as it's obvious I need to work things out a little more.

File: Configurations/descrip.mms.tmpl [hunk 2]
```diff
@@ -42,8 +42,18 @@
 
   # Because we need to make two computations of these data,
   # we store them in arrays for reuse
-  our @shlibs = map { $unified_info{sharednames}->{$_} || () } @{$unified_info{libraries}};
-  our @install_shlibs = map { $unified_info{sharednames}->{$_} || () } @{$unified_info{install}->{libraries}};
+  our @libs =
+      map { (my $x = $_) =~ s/\.a$//; $x }
+      grep(!/\.a$/, @{$unified_info{libraries}});
```
Comment (levitte) [inline Configurations/descrip.mms.tmpl]: Hmmm, there's something very odd with this...

File: Configurations/descrip.mms.tmpl [hunk 3]
```diff
@@ -42,8 +42,18 @@
...
+  our @shlibs =
+      map { $unified_info{sharednames}->{$_} || () }
+      grep(!/\.a$/, @{$unified_info{libraries}});
+  our @install_libs =
+      map { (my $x = $_) =~ s/\.a$//; $x }
+      grep(!/\.a$/, @{$unified_info{install}->{libraries}});
```
Comment (levitte) [inline Configurations/descrip.mms.tmpl]: ... same here

**Comment by levitte** [comment]
Ah, my private Appveyor build shows me that I finally got things right...

**Comment by levitte** [comment]
This is ready for review.

**Comment by levitte** [comment]
Ping?

**Comment by richsalz** [comment]
I'm looking at it now, but in the interim, how about the last of the code-health test program PR's? :)

File: include/openssl/crypto.h [hunk 4]
```diff
@@ -314,6 +314,8 @@ void CRYPTO_mem_debug_realloc(void *addr1, void *addr2, size_t num, int flag,
 void CRYPTO_mem_debug_free(void *addr, int flag,
         const char *file, int line);
 
+int CRYPTO_mem_leaks_cb(int (*cb) (const char *str, size_t len, void *u),
```
Comment (richsalz) [inline include/openssl/crypto.h]: Gotta document this (man3/OPENSSL_malloc.pod) ?  And does this break the existing functions which are documented?

File: test/bioprinttest.c [hunk 5]
```diff
@@ -247,3 +253,54 @@ int test_main(int argc, char **argv)
 
     return run_tests(argv[0]);
 }
+
+/*
+ * Replace testutil output routines.  We do this to eliminate possible sources
+ * if BIO error
```
Comment (richsalz) [inline test/bioprinttest.c]: "of BIO error"

File: test/bioprinttest.c [hunk 6]
```diff
@@ -247,3 +253,54 @@ int test_main(int argc, char **argv)
...
+{
+    return write(1, str, strlen(str));
+}
+
+int test_puts_stderr(const char *str)
+{
+    return write(2, str, strlen(str));
+}
+
+static char vprint_buf[10240];
```
Comment (richsalz) [inline test/bioprinttest.c]: optional: move this up to the top?
not optional: blank line after.

File: test/bioprinttest.c [hunk 7]
```diff
@@ -247,3 +253,54 @@ int test_main(int argc, char **argv)
...
+int test_vprintf_stdout(const char *fmt, va_list ap)
+{
+    size_t len = vsnprintf(vprint_buf, sizeof(vprint_buf), fmt, ap);
+
+    if (len >= sizeof(vprint_buf))
```
Comment (richsalz) [inline test/bioprinttest.c]: As I am found of saying in Matt's TLS1.3 PR's, you can use the ?: op here.

File: test/md2test.c [hunk 8]
```diff
@@ -10,7 +10,6 @@
 #include <string.h>
 
 #include "../e_os.h"
```
Comment (richsalz) [inline test/md2test.c]: should we just add -I.. to all flags and fix e_os.h include's while you're here touching everything?

File: test/siphash_internal_test.c [hunk 9]
```diff
@@ -226,7 +225,7 @@ static int test_siphash(int idx)
 
     if (expectedlen != SIPHASH_MIN_DIGEST_SIZE &&
         expectedlen != SIPHASH_MAX_DIGEST_SIZE) {
-        TEST_info("size %" OSSLzu " vs %d and %d", expectedlen,
+        TEST_info("size %zu vs %d and %d", expectedlen,
```
Comment (richsalz) [inline test/siphash_internal_test.c]: so much cleaner...

File: test/testutil.h [hunk 10]
```diff
@@ -335,3 +350,22 @@ void test_info_c90(const char *desc, ...) PRINTF_FORMAT(1, 2);
         }                                       \
     } while (0)
 #endif                          /* HEADER_TESTUTIL_H */
+
+
+/*
+ * The basic I/O functions used by the test framework.  These can be
+ * shimed when needed.  If shimed, they must ALL be replaced.
```
Comment (richsalz) [inline test/testutil.h]: shimmed.
Or maybe "overriden when needed.  Note that if one is, then all must be."

File: test/testutil/basic_output.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+/* Generate an error for anyone who tries to actually use this dummy */
+# define Applink_dummy "DON'T USE THIS"
+#endif
+
+void test_open_streams(void)
+{
+    bio_out = BIO_new_fd(1, 0);
+    bio_err = BIO_new_fd(2, 0);
+
+    OPENSSL_assert(bio_out != NULL && bio_err != NULL);
```
Comment (richsalz) [inline test/testutil/basic_output.c]: I have found separate asserts for separate conditions to be useful.

File: test/testutil/driver.c [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+static int num_test_cases = 0;
+
+void add_test(const char *test_case_name, int (*test_fn) ())
+{
+    assert(num_tests != OSSL_NELEM(all_tests));
+    all_tests[num_tests].test_case_name = test_case_name;
+    all_tests[num_tests].test_fn = test_fn;
+    all_tests[num_tests].num = -1;
+    ++num_test_cases;
+    ++num_tests;
```
Comment (richsalz) [inline test/testutil/driver.c]: flip these two increment lines.

File: test/testutil/driver.c [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+{
+    if (success)
+        ERR_clear_error();
+    else
+        ERR_print_errors_cb(err_cb, NULL);
+}
+
+static void helper_printf_stdout(const char *fmt, ...)
+{
+    va_list ap;
```
Comment (richsalz) [inline test/testutil/driver.c]: blank line after declaration

File: test/testutil/driver.c [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    va_start(ap, fmt);
+    test_vprintf_stdout(fmt, ap);
+    va_end(ap);
+}
+
+int run_tests(const char *test_prog_name)
+{
+    int num_failed = 0;
+
```
Comment (richsalz) [inline test/testutil/driver.c]: recycle/delete this blank line.

File: test/testutil/tests.c [hunk 15]
```diff
@@ -182,21 +44,30 @@ static void test_fail_message(const char *prefix, const char *file, int line,
                               const char *type, const char *fmt, ...)
             PRINTF_FORMAT(5, 6);
 
+static void helper_printf_stderr(const char *fmt, ...)
+{
+    va_list ap;
```
Comment (richsalz) [inline test/testutil/tests.c]: blank line after decl.

File: test/testutil/tests.c [hunk 16]
```diff
@@ -293,7 +164,7 @@ DEFINE_COMPARISONS(char, char, "%c")
 DEFINE_COMPARISONS(unsigned char, uchar, "%u")
 DEFINE_COMPARISONS(long, long, "%ld")
 DEFINE_COMPARISONS(unsigned long, ulong, "%lu")
-DEFINE_COMPARISONS(size_t, size_t, "%" OSSLzu)
+DEFINE_COMPARISONS(size_t, size_t, "%zu")
```
Comment (richsalz) [inline test/testutil/tests.c]: this doesn't need the same "not universally supported format" ifdef from the header file above?

**Review by richsalz — CHANGES_REQUESTED** [review]

File: Configurations/common.tmpl [hunk 17]
```diff
@@ -33,13 +34,22 @@
...
+     my %replace = ();
      while (@list) {
          my $item = shift @list;
          my $extensionlessitem = extensionlesslib($item);
-         push @newlist, $item
-             unless grep { $extensionlessitem eq extensionlesslib($_) } @list;
+         if (grep { $extensionlessitem eq extensionlesslib($_) } @list) {
+             if ($item ne $extensionlessitem) {
+                 # If this instance of the library is explicitely static, we
+                 # prefer that to any shared library name, since it must have
```
Comment (levitte) [inline Configurations/common.tmpl]: Fixing.

File: Configurations/common.tmpl [hunk 18]
```diff
@@ -33,13 +34,22 @@
...
+                 # the expliciteness must be done on purpose.
+                 $replace{$extensionlessitem} = $item;
+             }
+         } else {
+             push @newlist, $item;
+         }
      }
-     @newlist;
+     map { $replace{$_} // $_; } @newlist;
```
Comment (levitte) [inline Configurations/common.tmpl]: Yeah, I found it fairly recently myself :-)

File: crypto/mem_dbg.c [hunk 19]
```diff
@@ -443,7 +443,8 @@ void CRYPTO_mem_debug_realloc(void *addr1, void *addr2, size_t num,
 }
 
 typedef struct mem_leak_st {
-    BIO *bio;
+    int (*print_cb) (const char *str, size_t len, void *u);
```
Comment (levitte) [inline crypto/mem_dbg.c]: I modeled this after `ERR_print_errors_cb`...  And quite frankly, since this is getting into the public API, I'd prefer to keep the length parameter, just in case we *do* get to need it some day.  It doesn't hurt.

File: include/openssl/crypto.h (hunk 4 above)
Comment (levitte) [inline include/openssl/crypto.h]: It doesn't break the already existing ones at all.  But good point, I'll add the docs

File: crypto/mem_dbg.c (hunk 19 above)
Comment (richsalz) [inline crypto/mem_dbg.c]: okay.  And then you definitely should do a typedef for the callback.

File: test/bioprinttest.c (hunk 6 above)
Comment (levitte) [inline test/bioprinttest.c]: 1. Why move it far from where it's used?  Glad you made that optional ;-)
2. Sure thing.

File: test/bioprinttest.c (hunk 7 above)
Comment (levitte) [inline test/bioprinttest.c]: Sure, but I'd rather not.  I find it less readable for this kind of thing.

File: test/md2test.c (hunk 8 above)
Comment (levitte) [inline test/md2test.c]: I'm much more inclined to do it the other way around, actually...  Not sure if we have a team concensus for this.

File: test/testutil.h (hunk 10 above)
Comment (levitte) [inline test/testutil.h]: Yeah, "overriden" is probably better...

File: test/testutil/basic_output.c (hunk 11 above)
Comment (levitte) [inline test/testutil/basic_output.c]: Good point.  Fixing.

File: test/testutil/driver.c (hunk 12 above)
Comment (levitte) [inline test/testutil/driver.c]: Uhmmm, sure.  Not sure I see why, but sure.

File: test/testutil/driver.c (hunk 13 above)
Comment (levitte) [inline test/testutil/driver.c]: Yup

File: test/testutil/driver.c (hunk 14 above)
Comment (levitte) [inline test/testutil/driver.c]: Yup

File: test/testutil/tests.c (hunk 15 above)
Comment (levitte) [inline test/testutil/tests.c]: Done in all places I found in this file.

File: test/testutil/tests.c (hunk 16 above)
Comment (levitte) [inline test/testutil/tests.c]: Nope.  The condition in testutil.h is defined locally in each test program (currently, only bioprintttest.c), so it doesn't affect the implementation.  The reason to do this is that %zu isn't necessarely available with `vsnprintf()`

File: test/testutil/driver.c (hunk 12 above)
Comment (richsalz) [inline test/testutil/driver.c]: because it exactly parallels the function that follows.

**Comment by levitte** [comment]
Fixups committed.  I also rebased on a fresher master

**Review by richsalz — APPROVED** [review]

**Comment by levitte** [comment]
`bio_enc_test.c` slipped in and needed adapting as well...  re-confirm?

**Comment by richsalz** [comment]
reconfirm

**Comment by levitte** [comment]
Merged.  331058180aba047ac6b25da781eebc08b7267d2f to f044cd05a1d28c998731108dfcd25ba580cdda07

File: Configurations/common.tmpl [hunk 20]
```diff
@@ -113,7 +123,7 @@
  sub dolib {
      my $lib = shift;
      return "" if $cache{$lib};
-     unless ($disabled{shared}) {
+     unless ($disabled{shared} || $lib =~ /\.a$/) {
```
Comment (kaduk) [inline Configurations/common.tmpl]: (to pick just one instance) we're okay hardcoding ".a"?  (I guess the windows .lib is not exactly a direct analogue, but is close enough to think about.)

Comment (levitte) [inline Configurations/common.tmpl]: It's just an internal naming convention, and we already use `.o` as a similar convention.  The build files (Makefile et al) then know to convert them to their local convention.

I modeled the internal convention early on the normal Unix extensions because it's shorter to type and, well, I wasn't very imaginative.  I could have chosen `.object` and `.archive` just as well...
